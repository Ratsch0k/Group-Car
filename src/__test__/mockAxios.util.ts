jest.mock('axios');
import axios from 'axios';
const mockedAxios = axios as jest.Mocked<typeof axios>;

mockedAxios.head.mockImplementationOnce((path: string) => {
  if (path === '/auth') {
    return Promise.resolve({
      headers: {
        'xsrf-token': 'TEST TOKEN',
      },
    });
  }
});

export default mockedAxios;
