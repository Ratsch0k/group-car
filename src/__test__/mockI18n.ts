jest.mock('react-i18next', () => ({
  useTranslation: () => {
    return {
      t: (str: string) => str,
      i18n: {
        changeLanguage: () => new Promise(() => undefined),
        languages: ['de', 'en'],
        language: 'en',
      },
    };
  },
}));
