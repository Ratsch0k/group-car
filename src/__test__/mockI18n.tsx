import React from 'react';

jest.mock('react-i18next', () => ({
  useTranslation: () => {
    return {
      t: (str: string) => str,
      i18n: {
        changeLanguage: () => new Promise(() => undefined),
        languages: ['de', 'en'],
        language: 'en',
      },
    };
  },
  Trans: (props: any) => {
    return <div>{props.i18nKey}</div>;
  },
}));
