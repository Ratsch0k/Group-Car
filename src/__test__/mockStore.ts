/* eslint-disable @typescript-eslint/ban-types */
import {RootState} from '../lib/redux/store';
import thunk, {ThunkDispatch} from 'redux-thunk';
import {
  initialState as authState,
} from '../lib/redux/slices/auth/authSlice';
import {
  initialState as groupState,
} from '../lib/redux/slices/group/groupSlice';
import {
  initialState as invitesState,
} from '../lib/redux/slices/invites/invitesSlice';
import configureMockStore, {
  MockStoreCreator,
  MockStoreEnhanced,
} from 'redux-mock-store';
import {AnyAction, Middleware} from '@reduxjs/toolkit';

const defaultState: RootState = {
  router: {
    action: 'POP',
    location: {
      pathname: '/',
      search: '',
      hash: '',
      query: {},
      state: undefined,
    },
  },
  auth: authState,
  group: groupState,
  invites: invitesState,
};

type DispatchExts = ThunkDispatch<RootState, undefined, AnyAction>;


/**
 * Creates a mock store.
 * @param state Initial state. Will be merged with the default initial state.
 * @returns The mocked store
 */
function mockStore(state?: Partial<RootState>):
MockStoreEnhanced<RootState, DispatchExts> {
  const middlewares: Array<Middleware> = [thunk];
  const mockStoreCreator: MockStoreCreator<RootState, DispatchExts> =
  configureMockStore<RootState, DispatchExts>(middlewares);
  const initialState: RootState = {
    ...defaultState,
    ...state,
  };
  const store = mockStoreCreator(initialState);

  return store;
}

export default mockStore;
