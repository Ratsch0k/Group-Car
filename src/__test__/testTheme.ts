import {createMuiTheme} from '@material-ui/core';
import {themeProperties, IThemeOptions} from '../lib/theme';

export default createMuiTheme({
  ...themeProperties,
  props: {
    MuiButtonBase: {
      disableTouchRipple: true,
      disableRipple: true,
    },
  },
} as IThemeOptions);
