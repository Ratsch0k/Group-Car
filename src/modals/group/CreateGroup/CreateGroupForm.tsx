import React from 'react';
import {Grid} from '@material-ui/core';
import {FormTextField, ProgressButton} from 'lib';
import {useTranslation} from 'react-i18next';
import {useFormik} from 'formik';
import * as yup from 'yup';
import useGroups from 'lib/hooks/useGroups';

const minNameLength = 4;
const maxNameLength = 30;
const maxDescriptionLength = 200;

interface CreateGroupFormProps {
  setLoading?(value: boolean): void;
  close?(): void;
  navigateToManagement?(id: number): void;
}

/**
 * Form of the create group dialog.
 */
export const CreateGroupForm: React.FC<CreateGroupFormProps> =
(props: CreateGroupFormProps) => {
  const {t} = useTranslation();
  const {createGroup} = useGroups();

  const validationSchema = yup.object({
    name: yup.string().required(t('form.error.required'))
        .min(
            minNameLength,
            t(
                'form.error.tooShort',
                {min: minNameLength},
            ),
        )
        .max(
            maxNameLength,
            t(
                'form.error.tooLong',
                {max: maxNameLength},
            ),
        ),
    description: yup.string()
        .max(
            maxDescriptionLength,
            t(
                'form.error.tooLong',
                {max: maxDescriptionLength},
            ),
        ),
  });

  const formik = useFormik({
    initialValues: {
      name: '',
      description: undefined,
    },
    validationSchema,
    onSubmit: (values) => {
      props.setLoading && props.setLoading(true);
      createGroup(values.name, values.description).then((res) => {
        formik.setSubmitting(false);
        props.navigateToManagement && props.navigateToManagement(res.data.id);
      }).catch(() => {
        formik.setSubmitting(false);
        props.setLoading && props.setLoading(false);
      });
    },
  });

  return (
    <form onSubmit={formik.handleSubmit}>
      <Grid
        container
        direction='column'
        spacing={1}
      >
        <Grid item>
          <FormTextField
            id='create-group-name'
            label={t('modals.group.create.name')}
            name='name'
            formik={formik}
          />
        </Grid>
        <Grid item>
          <FormTextField
            id='create-group-description'
            label={t('modals.group.create.description')}
            multiline
            name='description'
            formik={formik}
            rows={6}
          />
        </Grid>
        <Grid item>
          <ProgressButton
            fullWidth
            id='create-group-submit'
            loading={formik.isSubmitting}
            variant='contained'
            type='submit'
            color='primary'
          >
            {t('modals.group.create.create')}
          </ProgressButton>
        </Grid>
      </Grid>
    </form>
  );
};

export default CreateGroupForm;
