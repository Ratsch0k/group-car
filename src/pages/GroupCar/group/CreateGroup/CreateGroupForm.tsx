import React from 'react';
import {Grid} from '@material-ui/core';
import {FormTextField, ProgressButton} from 'lib';
import {useTranslation} from 'react-i18next';
import {useFormik} from 'formik';
import * as yup from 'yup';
import {useAppDispatch} from 'lib/redux/hooks';
import {unwrapResult} from '@reduxjs/toolkit';
import {createGroup} from 'lib/redux/slices/group/groupThunks';

const minNameLength = 4;
const maxNameLength = 30;
const maxDescriptionLength = 200;

interface CreateGroupFormProps {
  setLoading?(value: boolean): void;
  close?(): void;
  navigateToManagement?(id: number): void;
}

/**
 * Form of the create group dialog.
 */
export const CreateGroupForm: React.FC<CreateGroupFormProps> =
(props: CreateGroupFormProps) => {
  const {t} = useTranslation();
  const dispatch = useAppDispatch();

  const validationSchema = yup.object({
    name: yup.string().required(t('form.error.required'))
      .min(
        minNameLength,
        t(
          'form.error.tooShort',
          {min: minNameLength},
        ),
      )
      .max(
        maxNameLength,
        t(
          'form.error.tooLong',
          {max: maxNameLength},
        ),
      ),
    description: yup.string()
      .max(
        maxDescriptionLength,
        t(
          'form.error.tooLong',
          {max: maxDescriptionLength},
        ),
      ),
  });

  const formik = useFormik({
    initialValues: {
      name: '',
      description: undefined,
    },
    validationSchema,
    onSubmit: async ({name, description}) => {
      props.setLoading && props.setLoading(true);

      try {
        const group = unwrapResult(await dispatch(
          createGroup({name, description}),
        ));

        formik.setSubmitting(false);
        props.navigateToManagement && props.navigateToManagement(group.id);
      } catch {
        formik.setSubmitting(false);
        props.setLoading && props.setLoading(false);
      }
    },
  });

  return (
    <form onSubmit={formik.handleSubmit}>
      <Grid
        container
        direction='column'
        spacing={1}
      >
        <Grid item>
          <FormTextField
            id='create-group-name'
            label={t('modals.group.create.name')}
            name='name'
            formik={formik}
            autoFocus
          />
        </Grid>
        <Grid item>
          <FormTextField
            id='create-group-description'
            label={t('modals.group.create.description')}
            multiline
            name='description'
            formik={formik}
            rows={6}
          />
        </Grid>
        <Grid item container justifyContent='flex-end'>
          <Grid item>
            <ProgressButton
              id='create-group-submit'
              loading={formik.isSubmitting}
              variant='contained'
              type='submit'
              color='primary'
              shadow
            >
              {t('modals.group.create.create')}
            </ProgressButton>
          </Grid>
        </Grid>
      </Grid>
    </form>
  );
};

export default CreateGroupForm;
