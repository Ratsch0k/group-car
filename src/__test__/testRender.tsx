/* eslint-disable @typescript-eslint/no-explicit-any */
import './mockAxios';
import './mockI18n';
import React from 'react';
import {render, RenderResult} from '@testing-library/react';
import {ConnectedRouter} from 'connected-react-router';
import {Provider} from 'react-redux';
import history from '../lib/redux/history';
import {StylesProvider} from '@material-ui/styles';
import {MuiThemeProvider} from '@material-ui/core';
import testTheme from './testTheme';
import {
  MockStoreEnhanced,
} from 'redux-mock-store';
import {RootState} from '../lib/redux/store';
import mockStore from './mockStore';
import mediaQuery from 'css-mediaquery';

export interface GlobalMocks {
  ResizeObserver: jest.Mock<{observe: jest.Mock, disconnect: jest.Mock}>;
}


export interface TestRenderResult {
  store: MockStoreEnhanced<unknown>;
  globalMocks: GlobalMocks;
}

/**
 * Testing generateClassName function to
 * avoid weird fails due to random class names
 * @param rule Rule
 * @param stylesheet Stylesheet
 * @returns The class name
 */
const generateClassName = (rule: any, stylesheet: any) => {
  return `${stylesheet.options.classNamePrefix}-${rule.key}`;
};

export interface TestRenderOptions {
  width?: string;
}

/**
 * Creates a match media to fix useMediaQuery in tests.
 * @param width Width of window
 * @returns Match media mock
 */
function createMatchMedia(width: any) {
  return (query: any) => ({
    matches: mediaQuery.match(query, {width}),
    addListener: () => undefined,
    removeEventListener: () => undefined,
  });
}


/**
 * Mock all global function which are used.
 */
const mockGlobals = (width: number): GlobalMocks => {
  // Mock ResizeObserver
  const resizeObserverMock = jest.fn().mockImplementation(() => ({
    observe: jest.fn(),
    disconnect: jest.fn(),
  }));
  window.ResizeObserver = resizeObserverMock;
  window.innerWidth = width;
  window.matchMedia = createMatchMedia(window.innerWidth) as any;

  return {ResizeObserver: resizeObserverMock};
};

/**
 * Render function for testing purposes.
 * Provides the necessary base components and
 * contexts.
 * @param state Redux state
 * @param children The children to render
 * @param options Additional options
 * @returns The result of the render function.
 */
function testRender(
  state: Partial<RootState>,
  children: React.ReactNode,
  options?: TestRenderOptions,
): RenderResult & TestRenderResult {
  const store = mockStore(state);
  const theme = testTheme(options && options.width);
  const width = theme.breakpoints.values[
    options && options.width ? options.width : 'lg'];
  const globalMocks = mockGlobals(width);

  const renderResult = render(
    <Provider store={store}>
      <ConnectedRouter history={history}>
        <MuiThemeProvider theme={theme}>
          <StylesProvider generateClassName={generateClassName}>
            {children}
          </StylesProvider>
        </MuiThemeProvider>
      </ConnectedRouter>
    </Provider>,
  );

  return {
    ...renderResult,
    globalMocks,
    store,
  };
}

export default testRender;
