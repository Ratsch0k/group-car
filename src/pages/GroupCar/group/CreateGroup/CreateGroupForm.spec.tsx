import  '../../../../__test__/mockAxios';
import '../../../../__test__/mockI18n';
import CreateGroupForm from "./CreateGroupForm";
import React from "react";
import testRender from "../../../../__test__/testRender";

it('renders without crashing', () => {
  testRender(
    {},
    <CreateGroupForm />
  );
})

it('matches snapshot', () => {
  const {baseElement} = testRender(
    {},
    <CreateGroupForm />,
  );

  expect(baseElement).toMatchSnapshot();
});