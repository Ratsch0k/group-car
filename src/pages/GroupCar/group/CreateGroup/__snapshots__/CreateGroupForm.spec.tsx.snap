// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<body>
  <div>
    <form>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-direction-xs-column"
      >
        <div
          class="MuiGrid-root MuiGrid-item"
        >
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            formik="[object Object]"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-marginDense MuiInputLabel-outlined MuiFormLabel-focused MuiInputLabel-focused"
              data-shrink="true"
              for="create-group-name"
              id="create-group-name-label"
            >
              modals.group.create.name
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-focused MuiOutlinedInput-focused MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
            >
              <input
                aria-describedby="create-group-name-helper-text"
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
                id="create-group-name"
                name="name"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="PrivateNotchedOutline-legendLabelled PrivateNotchedOutline-legendNotched"
                >
                  <span>
                    modals.group.create.name
                  </span>
                </legend>
              </fieldset>
            </div>
            <p
              class="MuiFormHelperText-root MuiFormHelperText-contained MuiFormHelperText-focused MuiFormHelperText-marginDense"
              id="create-group-name-helper-text"
            >
              <span>
                ​
              </span>
            </p>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item"
        >
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            formik="[object Object]"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined"
              data-shrink="false"
              for="create-group-description"
              id="create-group-description-label"
            >
              modals.group.create.description
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline MuiOutlinedInput-multiline MuiInputBase-marginDense MuiOutlinedInput-marginDense"
            >
              <textarea
                aria-describedby="create-group-description-helper-text"
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
                id="create-group-description"
                name="description"
                rows="6"
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="PrivateNotchedOutline-legendLabelled"
                >
                  <span>
                    modals.group.create.description
                  </span>
                </legend>
              </fieldset>
            </div>
            <p
              class="MuiFormHelperText-root MuiFormHelperText-contained MuiFormHelperText-marginDense"
              id="create-group-description-helper-text"
            >
              <span>
                ​
              </span>
            </p>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-justify-content-xs-flex-end"
        >
          <div
            class="MuiGrid-root MuiGrid-item"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-contained makeStyles-shadowPrimary makeStyles-button MuiButton-containedPrimary"
              id="create-group-submit"
              tabindex="0"
              type="submit"
            >
              <span
                class="MuiButton-label makeStyles-label"
              >
                modals.group.create.create
              </span>
            </button>
          </div>
        </div>
      </div>
    </form>
  </div>
</body>
`;
